evm/db/journal.py

```diff
@@ -1,226 +1,258 @@
         """
-        checkpoint_id = uuid.uuid4()
-        self.checkpoints.append(checkpoint_id)
-        self.journal_data[checkpoint_id] = {}
-        return checkpoint_id
+        changeset_id = uuid.uuid4()
+        self.journal_data[changeset_id] = {}
+        return changeset_id
 
-    def pop_checkpoint(self, checkpoint_id):
+    def pop_changeset(self, changeset_id: uuid.UUID) -> Dict[bytes, bytes]:
         """
-        Returns all changes from the given checkpoint.  This includes all of
-        the changes from any subsequent checkpoints, giving precidence to
-        earlier checkpoints.
+        Returns all changes from the given changeset.  This includes all of
+        the changes from any subsequent changeset, giving precidence to
+        later changesets.
         """
-        idx = self.checkpoints.index(checkpoint_id)
+        if changeset_id not in self.journal_data:
+            raise KeyError("Unknown changeset: {0}".format(changeset_id))
 
-        # update the checkpoint list
-        checkpoint_ids = self.checkpoints[idx:]
-        self.checkpoints = self.checkpoints[:idx]
+        all_ids = tuple(self.journal_data.keys())
+        changeset_idx = all_ids.index(changeset_id)
+        changesets_to_pop = all_ids[changeset_idx:]
 
-        # we pull all of the checkpoints *after* the checkpoint we are
-        # reverting to and collapse them to a single set of keys that need to
-        # be reverted (giving precidence to earlier checkpoints).
-        revert_data = merge(*(
+        # we pull all of the changesets *after* the changeset we are
+        # reverting to and collapse them to a single set of keys (giving
+        # precedence to later changesets)
+        changeset_data = merge(*(
             self.journal_data.pop(c_id)
             for c_id
-            in reversed(checkpoint_ids)
+            in changesets_to_pop
         ))
 
-        return dict(revert_data.items())
+        return changeset_data
 
-    def commit_checkpoint(self, checkpoint_id):
+    def commit_changeset(self, changeset_id: uuid.UUID) -> Dict[bytes, bytes]:
         """
-        Collapses all changes for the givent checkpoint into the previous
-        checkpoint if it exists.
+        Collapses all changes for the given changeset into the previous
+        changesets if it exists.
         """
-        changes_to_merge = self.pop_checkpoint(checkpoint_id)
-        if self.checkpoints:
-            # we only have to merge the changes into the latest checkpoint if
+        changeset_data = self.pop_changeset(changeset_id)
+        if not self.is_empty():
+            # we only have to merge the changes into the latest changeset if
             # there is one.
             self.latest = merge(
-                changes_to_merge,
                 self.latest,
+                changeset_data,
             )
+        return changeset_data
+
+    #
+    # Database API
+    #
+    def get(self, key: bytes) -> bytes:
```

Comment:
This should be:

```python
def get(self, key: bytes) -> Union[bytes, None]
    ...
```

right?

Reply:
If I'm not mistaken then every API in Python that accepts `T` implicitly accepts `Union[T, None]` and every API in Python that returns `T` implicitly possibly returns `Union[T, None]` hence making this extra type annotation would be superfluous. I'm think it would be non-standard to explicitly add this.